File: FeatureStorm/TwoPSet/Evals/AdHoc_Clarence.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
	"sync"
)

var mu sync.Mutex

type Set struct {
	id      int
	values  map[int]struct{}
	updates [][]string
}

func NewSet(id int) *Set {
	//fmt.Println("after creating ", id)
	createFile(id)
	//go c.check()
	return &Set{id: id, values: make(map[int]struct{})}
}

func (s *Set) Id() int {
	return s.id
}

func (s *Set) Values() map[int]struct{} {
	return s.values
}

func (s *Set) Contain(value int) bool {
	_, c := s.values[value]
	return c
}

func (s *Set) Add(value int) {
	mu.Lock()
	if !s.Contain(value) {
		s.values[value] = struct{}{}
	}
	mu.Unlock()
}

func (s *Set) Remove(value int) {
	mu.Lock()
	if s.Contain(value) {
		delete(s.values, value)
	}
	mu.Unlock()
}

func (s *Set) SetVal(opt_name string, value int) {
	if opt_name == "Add" {
		s.Add(value)
	} else if opt_name == "Remove" {
		s.Remove(value)
	}
	cur_upte := []string{opt_name, strconv.Itoa(value)}
	s.updates = append(s.updates, cur_upte)

	if opt_name == "Add" {
		rejct_up := opt_name + "_" + strconv.Itoa(value)
		addhoc_reject_local_update = make(map[string]bool)
		addhoc_reject_local_update[rejct_up] = true

		// fmt.Println("reject_local_update")
		// fmt.Println(reject_local_update)
	}
	val := addhoc_keys4Persist(s.Values())
	record(s.Id(), val, opt_name, value)
}

func (s *Set) SetRemoteVal(rid int, opt_name string, value int) {
	if opt_name == "Add" {
		s.Add(value)
	} else if opt_name == "Remove" {
		s.Remove(value)
	}
}

func (s *Set) Union(o *Set) {

	mu.Lock()
	// fmt.Println("Starting to merge:")
	// s.Print()
	// o.Print()
	m := make(map[int]bool)

	for item := range s.values {
		m[item] = true
	}

	for item := range o.values {
		if _, ok := m[item]; !ok {
			s.values[item] = struct{}{}
		}
	}
	// fmt.Print("Merged ")
	// s.Print()
	mu.Unlock()
}

func (s *Set) Merge(rid int, r_updates [][]string) {
	// mu.Lock()
	// res := fmt.Sprintf("%s%d:%d", "Counter_", rid, rval)
	fmt.Println("Starting to merge req from replica_", rid)

	tmpSet := new(Set)
	tmpSet.id = rid
	tmpSet.values = make(map[int]struct{})
	if len(r_updates) > 0 {
		for i := 0; i < len(r_updates); i++ {
			req_val, _ := strconv.Atoi(r_updates[i][1])
			if r_updates[i][0] == "Add" {
				tmpSet.Add(req_val)
			} else {
				tmpSet.Remove(req_val)
			}
		}
	}

	// if len(r_updates) > 0 {
	// 	for i := 0; i < len(r_updates); i++ {
	// 		req_val, _ := strconv.Atoi(r_updates[i][1])
	// 		s.SetRemoteVal(rid, r_updates[i][0], req_val)
	// 	}
	// }

	s.Union(tmpSet)
	tmpSet = nil
	fmt.Print("Merged :-->")
	s.Print()
	// mu.Unlock()
}

func (s *Set) Print() {
	fmt.Print("Set:", s.id, " ")
	values := get_keys_from_map(s.values)
	fmt.Println(values)
}

func get_keys_from_map(mapp map[int]struct{}) []int {
	keys := make([]int, 0, len(mapp))
	for key := range mapp {
		keys = append(keys, key)
	}
	sort.Ints(keys[:])
	return keys
}

func (s *Set) ToMarshal() []byte {
	id_2d := []string{strconv.Itoa(s.Id()), ""}
	s.updates = append(s.updates, id_2d)

	jsonData, err := json.Marshal(s.updates)
	if err != nil {
		fmt.Println("Error while marshaling updates")
		return nil
	}
	s.updates = [][]string{}
	return jsonData
}

func FromMarshalData(bytes []byte) (int, [][]string) {

	var remote_updates [][]string
	err := json.Unmarshal(bytes, &remote_updates)

	if err != nil {
		fmt.Println("Error while unmarshaling ", err)
		return -1, nil
	}

	rid, _ := strconv.Atoi(remote_updates[len(remote_updates)-1][0])
	// fmt.Println("rid is: ", rid)
	if len(remote_updates) == 1 {
		return rid, nil
	}
	r_updates := remote_updates[0 : len(remote_updates)-1]
	// fmt.Println(r_updates)
	return rid, r_updates
}

// Wrapper's functionality
var addhoc_updates = 0
var addhoc_undo_opt_not_arrived string
var addhoc_reject_local_update map[string]bool

func addhoc_keys4Persist(mapp map[int]struct{}) string {
	keys := make([]int, 0, len(mapp))
	for key := range mapp {
		keys = append(keys, key)
	}
	sort.Ints(keys[:])

	jsonStr, _ := json.Marshal(keys)
	return string(jsonStr)
}

func (s *Set) addhoc_setRemoteWrap(rid int, opt_name string, value int) {
	// s.SetRemoteVal(rid, opt_name, value)
	// val := keys4Persist(s.Values())
	remoteRecord(s.Id(), rid, opt_name, value)
}

func (s *Set) addhoc_mergeWrap(rid int, r_updates [][]string) {
	if len(r_updates) > 0 {
		// fmt.Println(r_updates)
		if r_updates[0][0] == "undo" {
			fmt.Println("undo requested from replica:", rid)
			// fmt.Println("undo req from others")
			// fmt.Println(r_updates)
			undo_info := strings.Split(r_updates[0][1], "_")
			undo_val, _ := strconv.Atoi(undo_info[1])
			remote_Lt, _ := strconv.Atoi(undo_info[2])
			if undo(rid, undo_info[0], undo_val, remote_Lt) {
				fmt.Println("Undo executing")
				if undo_info[0] == "Add" {
					s.SetRemoteVal(rid, "Remove", undo_val)
					s.addhoc_setRemoteWrap(rid, "Remove", undo_val)
				} else if undo_info[0] == "Remove" {
					s.SetRemoteVal(rid, "Add", undo_val)
					s.addhoc_setRemoteWrap(rid, "Add", undo_val)
				}
				val := addhoc_keys4Persist(s.Values())
				fmt.Println("Set:", s.Id(), val)
				record(s.Id(), val, "Undo done from replica", rid)

				// Revisit the previous rejected local updates
				fmt.Println("Revisit previous updates")
				check_opt := undo_info[0] + "_" + undo_info[1]
				fmt.Println("revisit opt:", check_opt)

				if addhoc_reject_local_update[check_opt] == true {
					fmt.Println("Execute previous rejected local update due to undone now")
					s.SetVal(undo_info[0], undo_val)
					revist_val := addhoc_keys4Persist(s.Values())
					record(s.Id(), revist_val, undo_info[0]+"_revisit", undo_val)
				}

			} else {
				fmt.Println("Undo not required as this operation has not merged yet")
				fmt.Println("Recording this undo info for future rejection")
				addhoc_undo_opt_not_arrived += strconv.Itoa(rid) + "_" + undo_info[0] + "_" + undo_info[1]
				// fmt.Println("undo_opt_not_arrived:", undo_opt_not_arrived)
			}
		} else {
			// fmt.Println("req replica:", rid)
			// fmt.Println(r_updates)
			for i := 0; i < len(r_updates); i++ {

				check := strconv.Itoa(rid) + "_" + r_updates[i][0] + "_" + r_updates[i][1]
				// fmt.Println("check:", check)
				if check == addhoc_undo_opt_not_arrived {
					fmt.Println("No need to merge as already undo req by same replica")
				} else {
					req_val, _ := strconv.Atoi(r_updates[i][1])
					s.addhoc_setRemoteWrap(rid, r_updates[i][0], req_val)
				}
			}
			s.Merge(rid, r_updates)

			val := addhoc_keys4Persist(s.Values())
			record(s.Id(), val, "Merge from replica", rid)
		}
	}
}

// C++ portion
type LamportClock struct {
	time int
	mu   sync.Mutex
}

func (lc *LamportClock) localEvent() int {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	lc.time++
	return lc.time
}

func (lc *LamportClock) getTime() int {
	lc.mu.Lock()
	defer lc.mu.Unlock()
	return lc.time
}

var id2ver = make(map[int]int)
var updateList []Update
var cClock LamportClock
var rid2LT = make(map[int]*LamportClock)

type Update struct {
	ID    int
	Ver   int
	Value string
}

func createFile(id int) {
	dirPath := "DBs/"
	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
		err := os.Mkdir(dirPath, 0700)
		if err != nil {
			fmt.Println("Error creating directory:", err)
			return
		}
	}
	filename := dirPath + strconv.Itoa(id) + ".txt"
	file, err := os.Create(filename)
	if err != nil {
		fmt.Println("Error creating file:", err)
		return
	}
	fmt.Println("Storage file created for Replica", id)
	file.Close()

	id2ver[id] = 0
}

func record(id int, val string, optName string, upValue int) {
	// Write to physical file
	filename := "DBs/" + strconv.Itoa(id) + ".txt"
	id2ver[id]++
	file, err := os.OpenFile(filename, os.O_APPEND|os.O_WRONLY, 0600)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	logEntry := fmt.Sprintf("R_ID:%d Ver:%d St:%s LT:%d Opt:%s_%d Type:Local\n", id, id2ver[id], val, cClock.localEvent(), optName, upValue)
	if _, err := file.WriteString(logEntry); err != nil {
		fmt.Println("Error writing to file:", err)
		return
	}

	// Update list
	updateList = append(updateList, Update{id, id2ver[id], val})
}

func remoteRecord(id int, rid int, optName string, upValue int) {
	filename := "DBs/" + strconv.Itoa(id) + ".txt"
	file, err := os.OpenFile(filename, os.O_APPEND|os.O_WRONLY, 0600)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	remoteLT := rid2LT[rid].localEvent()
	logEntry := fmt.Sprintf("R_ID:%d Opt:%s_%d Type:{Remote, rid:%d local_LT:%d}\n", id, optName, upValue, rid, remoteLT)
	if _, err := file.WriteString(logEntry); err != nil {
		fmt.Println("Error writing to file:", err)
		return
	}
}

func undo(rID int, undoUpdate string, undoVal int, rLT int) bool {
	if rid2LT[rID].getTime() > rLT {
		return true
	}
	return false
}

File: Legion/applications/examples/Evals/AdHoc_Clarence.js
const fs = require('fs');
const path = require('path');

// Utility to format logs
function formatLog(id, val, optName, upValue, lamportTime, isRemote = false, rid = null) {
    let logEntry = `R_ID:${id} St:${val} LT:${lamportTime} Opt:${optName}_${upValue} `;
    if (isRemote) {
        logEntry += `Type:{Remote, rid:${rid} local_LT:${rid2LT[rid].localEvent()}}\n`;
    } else {
        logEntry += `Type:Local\n`;
    }
    return logEntry;
}

// Highly nested structure for version control and recording
const recordManager = {
    versionMap: {},
    recordUpdate(id, val, optName, upValue, isRemote, rid) {
        const lamportTime = cClock.localEvent();
        this.incrementVersion(id);
        const logEntry = formatLog(id, val, optName, upValue, lamportTime, isRemote, rid);
        this.writeLogToFile(id, logEntry);
        if (!isRemote) {
            updateList.push(new Update(id, id2ver[id], val));
        }
    },
    incrementVersion(id) {
        if (!this.versionMap[id]) {
            this.versionMap[id] = 0;
        }
        this.versionMap[id]++;
    },
    writeLogToFile(id, logEntry) {
        const filePath = path.join(__dirname, 'DBs', `${id}.txt`);
        fs.appendFileSync(filePath, logEntry);
    },
    getVersion(id) {
        return this.versionMap[id] || 0;
    }
};

// Separate class to handle JSON marshalling/unmarshalling
class JSONHandler {
    static marshal(data) {
        return JSON.stringify(data);
    }

    static unmarshal(jsonStr) {
        return JSON.parse(jsonStr);
    }
}

// Complex Set Manager which depends on several other classes
class SetManager {
    constructor() {
        this.sets = {};
    }

    createSet(id) {
        const newSet = new Set(id);
        this.sets[id] = newSet;
        return newSet;
    }

    getSet(id) {
        return this.sets[id];
    }

    handleSetOperation(id, operation, value) {
        const set = this.getSet(id);
        if (!set) throw new Error(`Set with id ${id} not found!`);
        set.SetVal(operation, value);
    }
}

class Set {
    constructor(id) {
        this.id = id;
        this.values = {};
        this.updates = [];
        this.addhoc_reject_local_update = {};
        createFile(id);
        this.recordManager = recordManager;
    }

    Id() {
        return this.id;
    }

    Values() {
        return this.values;
    }

    Contain(value) {
        return this.values.hasOwnProperty(value);
    }

    Add(value) {
        if (!this.Contain(value)) {
            this.values[value] = true;
        }
    }

    Remove(value) {
        if (this.Contain(value)) {
            delete this.values[value];
        }
    }

    SetVal(opt_name, value) {
        if (opt_name === 'Add') {
            this.Add(value);
        } else if (opt_name === 'Remove') {
            this.Remove(value);
        }
        const cur_upte = [opt_name, value.toString()];
        this.updates.push(cur_upte);

        if (opt_name === 'Add') {
            let rejct_up = `${opt_name}_${value}`;
            this.addhoc_reject_local_update[rejct_up] = true;
        }
        const val = addhoc_keys4Persist(this.Values());
        this.recordManager.recordUpdate(this.Id(), val, opt_name, value, false);
    }

    SetRemoteVal(rid, opt_name, value) {
        if (opt_name === 'Add') {
            this.Add(value);
        } else if (opt_name === 'Remove') {
            this.Remove(value);
        }
        this.recordManager.recordUpdate(this.Id(), this.Values(), opt_name, value, true, rid);
    }

    Union(o) {
        const m = {};
        Object.keys(this.values).forEach((item) => m[item] = true);
        Object.keys(o.values).forEach((item) => {
            if (!m.hasOwnProperty(item)) {
                this.values[item] = true;
            }
        });
    }

    Merge(rid, r_updates) {
        const tmpSet = new Set(rid);
        r_updates.forEach(([opt_name, val]) => {
            let req_val = parseInt(val, 10);
            if (opt_name === 'Add') {
                tmpSet.Add(req_val);
            } else {
                tmpSet.Remove(req_val);
            }
        });

        this.Union(tmpSet);
        this.Print();
    }

    Print() {
        const values = Object.keys(this.values).map(Number).sort((a, b) => a - b);
        console.log(`Set:${this.id}`, values);
        this.recordManager.recordUpdate(this.id, JSON.stringify(values), "Print", 0, false);
    }

    ToMarshal() {
        this.updates.push([this.Id().toString(), ""]);
        return JSONHandler.marshal(this.updates);
    }

    addhoc_setRemoteWrap(rid, opt_name, value) {
        this.recordManager.recordUpdate(this.Id(), JSONHandler.marshal(this.Values()), opt_name, value, true, rid);
    }

    addhoc_mergeWrap(rid, r_updates) {
        if (r_updates.length > 0 && r_updates[0][0] === 'undo') {
            let [undo_opt, undo_val, remote_Lt] = r_updates[0][1].split('_');
            if (undo(rid, undo_opt, parseInt(undo_val, 10), parseInt(remote_Lt, 10))) {
                if (undo_opt === 'Add') {
                    this.SetRemoteVal(rid, 'Remove', parseInt(undo_val, 10));
                    this.addhoc_setRemoteWrap(rid, 'Remove', parseInt(undo_val, 10));
                } else if (undo_opt === 'Remove') {
                    this.SetRemoteVal(rid, 'Add', parseInt(undo_val, 10));
                    this.addhoc_setRemoteWrap(rid, 'Add', parseInt(undo_val, 10));
                }
            }
        } else {
            this.Merge(rid, r_updates);
        }
    }
}

// Creates a file for storing logs related to a specific Set
function createFile(id) {
    const dirPath = path.join(__dirname, 'DBs');
    if (!fs.existsSync(dirPath)) {
        fs.mkdirSync(dirPath);
    }
    const filePath = path.join(dirPath, `${id}.txt`);
    fs.writeFileSync(filePath, "");
}

// Special Update class for record management
class Update {
    constructor(id, version, value) {
        this.ID = id;
        this.Version = version;
        this.Value = value;
    }
}

// Highly nested function that generates and persists keys
function addhoc_keys4Persist(values) {
    const sortedKeys = Object.keys(values).map(Number).sort((a, b) => a - b);
    const jsonStr = JSONHandler.marshal(sortedKeys);
    return jsonStr;
}

// Lamport Clock logic with complex dependency on Set and record keeping
class LamportClock {
    constructor() {
        this.time = 0;
        this.mu = false;  // Simulating a mutex lock
    }

    localEvent() {
        this.incrementTime();
        return this.time;
    }

    incrementTime() {
        if (!this.mu) {
            this.mu = true; // Lock
            this.time++;
            this.mu = false; // Unlock
        }
    }

    getTime() {
        return this.time;
    }
}

// Clock instances, update tracking, and record keeping
const cClock = new LamportClock();
const rid2LT = {};
const id2ver = {};
const updateList = [];

// Undo function that depends on multiple modules
function undo(rID, undoUpdate, undoVal, rLT) {
    if (rid2LT[rID].getTime() > rLT) {
        return true;
    }
    return false;
}
